File: Rado/part365/src/app/model/displayPartView.ts
import { StaticSelectionService } from "@services/staticSelection.service";
import { ItemType } from "./enum/itemType.enum";
import { ApprovedType } from "./enum/approveType.enum";
import { UserType } from "./enum/userType.enum";
import { ImageData } from "./imageData";
import { NgxGalleryImage } from "@app/ngx-gallery/models/ngx-gallery-image.model";
import { convertImage, dateString, isPart } from "@app/functions/functions";

export class DisplayPartView {
  id?               : number;
  itemType?         : ItemType;
  sellerName?       : string;
  sellerPhone?      : string;
  sellerPhone2?     : string;
  sellerViber?      : string;
  sellerWhats?      : string;
  sellerCity?       : string;
  sellerLogo?       : string;
  sellerWebPage?    : string;
  // extra
  ngImages?         : NgxGalleryImage[] = [];
  numberImages?     : number;
  mainImageData?    : ImageData;
  images?           : ImageData[];
  mainImage?        : NgxGalleryImage;

  userId?           : number;
  regionId?         : number;
  approved?         : ApprovedType
  mainPicture?      : string;

  isCar?            : boolean;
  bus?              : number;
  carId?            : number
  companyId?        : number;
  modelId?          : number;
  modificationId?   : number;
  companyName?      : string;
  modelName?        : string;
  modificationName? : string;
  dealerSubCategoryName?    : string;
  partNumber?       : string;
  price?            : number;
  year?             : number;
  description?      : string;
  vin?              : string;
  engineType?       : number;
  engineModel?      : string;
  gearboxType?      : number;
  millage?           : number;
  powerBHP?         : number;
  modifiedTime?     : number;
  leftRightPosition?: number;
  frontBackPosition?: number;
  categoryName?     : string;
  sellerCompanyName?: string;

  yearName?         : string;

  descriptionModel? : string = '';
  positionDesc?     : string;
  modifiedTimeDesc? : string;
  phonesDesc?       : string;
  engineTypeDesc?   : string;
  gearboxDesc?      : string;

  tyreCompanyId?    : number;
  tyreWidth?        : number;
  tyreHeight?       : number;
  tyreRadius?       : number;
  tyreType?         : number;
  rimWidth?         : number;
  rimMaterial?      : number;
  rimOffset?        : number;
  rimBoltCount?     : number;
  rimBoltDistance?  : number;
  rimCenter?        : number;

  tyreCompanyName?  : string;
  tyreWidthName?    : string;
  tyreHeightName?   : string;
  tyreRadiusName?   : string;
  tyreTypeName?     : string;
  rimWidthName?     : string;
  rimMaterialName?  : string;
  rimOffsetName?    : string;
  rimBoltCountName? : string;
  rimBoltDistanceName?: string;
  rimCenterName?    : string;
  count?            : number;
  monthDOT?         : number;
  yearDOT?          : number;

  tags?             : object;
  rimTags?          : object;
  tyreTags?         : object;
  partTagsMap?         : Map<string, string>;
  rimTagsMap?          : Map<string, string>;
  tyreTagsMap?         : Map<string, string>;
  regionName?       : string;
  dealer?           : UserType;
  isRim?:            boolean;
  isTyre?:        boolean;
}

function getPhones(data: DisplayPartView) {
  const _phones = [];
  if (data.sellerPhone) _phones.push(data.sellerPhone);
  if (data.sellerPhone2) _phones.push(data.sellerPhone2);
  if (_phones.length)
    return _phones.join('/');

  return undefined;
}

export function Enrich(data: DisplayPartView, staticService: StaticSelectionService): DisplayPartView {
  
//   if (data.part) {
//     if (data.isCar)
//       data.id = data.part.carId;
//     else 
//       data.id = data.part.partId;
//     data.bus = data.part.bus;
//     data.isCar = data.part.isCar;
//     data.carId = data.part.carId
//     data.companyId = data.part.companyId;
//     data.modelId = data.part.modelId;
//     data.modificationId = data.part.modificationId;
//     data.companyName = data.part.companyName
//     data.modelName = data.part.modelName
//     data.modificationName = data.part.modificationName
//     data.dealerSubCategoryName = data.part.dealerSubCategoryName;
//     data.partNumber = data.part.partNumber
//     data.price = data.part.price
//     data.year = data.part.year
//     data.vin = data.part.vin
//     data.engineType = data.part.engineType
//     data.engineModel = data.part.engineModel
//     data.gearboxType = data.part.gearboxType
//     data.millage = data.part.millage
//     data.powerBHP = data.part.powerBHP
//     data.modifiedTime = data.part.modifiedTime
//     data.leftRightPosition = data.part.leftRightPosition
//     data.frontBackPosition = data.part.frontBackPosition
//     data.categoryName = data.part.categoryName
//     data.yearName = data.part.yearName;
// }
//   if (data.rimWithTyre) {
//     data.companyId = data.rimWithTyre.companyId;
//     data.modelId = data.rimWithTyre.modelId;
//     data.companyName = data.rimWithTyre.companyName;
//     data.modelName = data.rimWithTyre.modelName;
//     data.tyreCompanyId = data.rimWithTyre.tyreCompanyId;
//     data.tyreWidth = data.rimWithTyre.tyreWidth;
//     data.tyreHeight = data.rimWithTyre.tyreHeight;
//     data.tyreRadius = data.rimWithTyre.tyreRadius;
//     data.tyreType = data.rimWithTyre.tyreType;
//     data.rimWidth = data.rimWithTyre.rimWidth;
//     data.rimMaterial = data.rimWithTyre.rimMaterial;
//     data.rimOffset = data.rimWithTyre.rimOffset;
//     data.rimBoltCount = data.rimWithTyre.rimBoltCount;
//     data.rimBoltDistance = data.rimWithTyre.rimBoltDistance;
//     data.rimCenter = data.rimWithTyre.rimCenter;
//     data.modifiedTime = data.rimWithTyre.modifiedTime
//     data.count = data.rimWithTyre.count;
//     data.monthDOT = data.rimWithTyre.monthDOT;
//     data.yearDOT = data.rimWithTyre.yearDOT;
//   }
  try {
    data.mainImage = convertImage(data.mainImageData!);
    const images_: NgxGalleryImage[] = []
    data.images?.forEach((image) => {
        const convertedImage = convertImage(image)
        if (convertedImage) {
            images_.push(convertedImage)
        }
    })
    data.ngImages = [...images_]  }
  catch
  {
    console.log(data)
  }
  data.regionName = staticService.getRegion(data.regionId!);
  data.modifiedTimeDesc = dateString(data.modifiedTime!);
  data.phonesDesc = getPhones(data);

  if (isPart(data.itemType!)) {
    data.engineTypeDesc = staticService.engineTypeDescription(data.engineType!);
    data.gearboxDesc = staticService.GearboxType.filter(x => x.value === data.gearboxType)[0]?.text;
  }
  data.tyreRadiusName = staticService.TyreRadius.find(elem => elem.value === data.tyreRadius)?.text;

  if (data.itemType === ItemType.Tyre || data.itemType === ItemType.RimWithTyre) {
    data.tyreCompanyName = staticService.TyreProducers.find(elem => elem.value === data.tyreCompanyId)?.text;
    data.tyreWidthName = staticService.TyreWidth.find(elem => elem.value === data.tyreWidth)?.text;
    data.tyreHeightName = staticService.TyreHeight.find(elem => elem.value === data.tyreHeight)?.text;
    data.tyreTypeName = staticService.TyreType.find(elem => elem.value === data.tyreType)?.text;
  }
  if (data.itemType === ItemType.Rim || data.itemType === ItemType.RimWithTyre) {
    data.rimWidthName = staticService.RimWidth.find(elem => elem.value === data.rimWidth)?.text;
    data.rimMaterialName = staticService.RimMaterial.find(elem => elem.value === data.rimMaterial)?.text;
    data.rimOffsetName = staticService.RimOffset.find(elem => elem.value === data.rimOffset)?.text;
    data.rimBoltCountName = data.rimBoltCount?.toString();
    data.rimBoltDistanceName = staticService.BoltDistance.find(elem => elem.value === data.rimBoltDistance)?.text;
    data.rimCenterName = staticService.RimCenter.find(elem => elem.value === data.rimCenter)?.text;
  }


  if (data.tags)
    data.partTagsMap = new Map(Object.entries(data.tags!));
  else 
    data.partTagsMap = new Map();

  if (data.rimTags)
    data.rimTagsMap = new Map(Object.entries(data.rimTags));
  else 
    data.rimTagsMap = new Map();

  if (data.tyreTags)
    data.tyreTagsMap = new Map(Object.entries(data.tyreTags));
  else 
    data.tyreTagsMap = new Map();

    data.isTyre = data.itemType == ItemType.Tyre || data.itemType == ItemType.RimWithTyre;
    data.isRim = data.itemType == ItemType.Rim || data.itemType == ItemType.RimWithTyre;

  // data.companyName = data.rimTags?.get("Компания");
  return data;
}

File: Models/Enums/DisplayPartView.cs
using Models.Helper;
using Models.Models;
using Rado.Enums;
using Rado.Models;
using System.Text.Json.Serialization;

namespace Models.Enums
{
    public class DisplayPartView
    {
        public long Id { get; set; }
        public ItemType ItemType { get; set; }
        public bool IsCar { get; set; }
        [JsonIgnore()]
        public PartView Part { get; set; }
        [JsonIgnore()]
        public RimWithTyreView RimWithTyre { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? MainPicture { get; set; }
        public string? Description { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public decimal Price { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public string? SellerName { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public string? SellerPhone { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public string? SellerPhone2 { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? SellerViber { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public string? SellerWhats { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public string? SellerLogo { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public string SellerCity { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public string? SellerCompanyName { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public string? SellerWebPage { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public int NumberImages { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public int UserId { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public ImageDataClass MainImageData { get; set; }
        [JsonIgnore()]
        public int RegionId { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public long MainImageId { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] 
        public ImageDataClass[] Images { get; set; }
        [JsonIgnore()]
        public int Approved { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public int Count { get; set; }
        public UserType Dealer { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public long ModifiedTime { get; set; }
        [JsonIgnore()]
        public string? EngineTypeDesc { get; set; }
        [JsonIgnore()]
        public string? PositionDesc { get; set; }
        [JsonIgnore()]
        public string? GearboxDesc { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? ItemTypeStr { get; set; }
        [JsonIgnore()]
        public string? RegionStr { get; set; }

        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? DescriptionModel { get; set; }

        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public Dictionary<string, string>? Tags { get; set; } = new Dictionary<string, string>();
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public Dictionary<string, string>? RimTags { get; set; } = new Dictionary<string, string>();
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public Dictionary<string, string>? TyreTags { get; set; } = new Dictionary<string, string>();
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? CompanyName { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? ModelName { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? ModificationName { get; set; }

        public DisplayPartView Normalize()
        {
            Tags = new Dictionary<string, string>();
            switch(ItemType)
            {
                case ItemType.CarPart:
                case ItemType.BusPart:
                    {
                        NormalizePart(); break;
                    }
                case ItemType.OnlyBus:
                case ItemType.OnlyCar:
                    {
                        NormalizeCar(); break;
                    }
                case ItemType.Tyre:
                    {
                        NormalizeTyre(); break; 
                    }
                case ItemType.Rim:
                    {
                        NormalizeRim(); break;
                    }
                case ItemType.RimWithTyre:
                    {
                        NormalizeRimWithTyre(); break;  
                    }
            }

            return this;
        }

        private string GetPosition() {
            Part.FrontBackPosition = Part.FrontBackPosition ?? 0;
            Part.LeftRightPosition = Part.LeftRightPosition ?? 0;

            if (Part.FrontBackPosition == (int)PositionFB.None && Part.LeftRightPosition == (int)PositionLR.None)
            {
                return "";
            }

            string positionDesc = "";

            if (Part.FrontBackPosition != 0)
            {
                positionDesc = ConverterToString.FrontBackStr((PositionFB)Part.FrontBackPosition);
            }

            if (Part.LeftRightPosition != 0)
            {
                positionDesc += ConverterToString.LeftRightStr((PositionLR)Part.LeftRightPosition);
            }

            return positionDesc;
        }

        private void NormalizePart()
        {
            NormalizePartCar();
            NormalzeCommon();
        }
        private void NormalizeCar()
        {
            NormalizePartCar();
            NormalzeCommon();
        }
        private void NormalizeTyre(bool only = true)
        {
            TyreTags = new Dictionary<string, string>();
            string tyreWidth = "", tyreHeight = "", tyreRadius = "";
            if (ConverterToString.TyreWidthString(RimWithTyre.TyreWidth) != string.Empty)
            {
                tyreWidth = ConverterToString.TyreWidthString(RimWithTyre.TyreWidth);
                TyreTags.Add("Широчина", tyreWidth);
            }

            if (ConverterToString.TyreHeightString(RimWithTyre.TyreHeight) != string.Empty)
            {
                tyreHeight = ConverterToString.TyreHeightString(RimWithTyre.TyreHeight);
                TyreTags.Add("Височина", tyreHeight);
            }

            if (ConverterToString.TyreRadiusString(RimWithTyre.TyreRadius) != string.Empty)
            {
                tyreRadius = ConverterToString.TyreRadiusString(RimWithTyre.TyreRadius);
                TyreTags.Add("Радиус", tyreRadius);
            }

            if (tyreWidth.Length > 0 && tyreHeight.Length > 0 && tyreRadius.Length > 0)
                TyreTags.Add("Размер", $"{tyreWidth}/{tyreHeight}R{tyreRadius}");

            if (ConverterToString.TyreTypeString(RimWithTyre.TyreType) != string.Empty)
                TyreTags.Add("Вид", ConverterToString.TyreTypeString(RimWithTyre.TyreType));

            if (RimWithTyre.MonthDOT != null && RimWithTyre.MonthDOT != 0 &&
                RimWithTyre.YearDOT != null && RimWithTyre.YearDOT != 0)
            {
                string year = RimWithTyre.YearDOT.Value.ToString("D2");
                string month = RimWithTyre.MonthDOT.Value.ToString("D2");
                TyreTags.Add("DOT", $"{month}{year}");
            }

            if (ConverterToString.TyreProducersString(RimWithTyre.TyreCompanyId) != string.Empty)
                TyreTags.Add("Производител", ConverterToString.TyreProducersString(RimWithTyre.TyreCompanyId));

            if (RimWithTyre.Count != 0)
                TyreTags.Add("Брой", $"{RimWithTyre.Count}");

            if (only)
                NormalzeCommon();
        }

        private void NormalizeRim(bool only = true)
        {
            RimTags = new Dictionary<string, string>();

            if (ConverterToString.RimWidthString(RimWithTyre.RimWidth) != string.Empty)
                RimTags.Add("Щирочина", ConverterToString.RimWidthString(RimWithTyre.RimWidth));

            if (ConverterToString.RimMaterialString(RimWithTyre.RimMaterial) != string.Empty)
                RimTags.Add("Материал", ConverterToString.RimMaterialString(RimWithTyre.RimMaterial));

            if (RimWithTyre.RimCenter != null && RimWithTyre.RimCenter.Value != 0 &&
                ConverterToString.RimCenterString(RimWithTyre.RimCenter.Value) != string.Empty)
                RimTags.Add("Център", ConverterToString.RimCenterString(RimWithTyre.RimCenter.Value));

            if (RimWithTyre.RimOffset != null && ConverterToString.RimOffsetString(RimWithTyre.RimOffset.Value) != string.Empty)
                RimTags.Add("Офсет", ConverterToString.RimOffsetString(RimWithTyre.RimOffset.Value));

            if (RimWithTyre.RimBoltCount != null && ConverterToString.RimBoltCountString(RimWithTyre.RimBoltCount.Value) != string.Empty)
                RimTags.Add("Брой болтове", ConverterToString.RimBoltCountString(RimWithTyre.RimBoltCount.Value));

            if (RimWithTyre.RimBoltDistance != null && ConverterToString.RimBoltDistanceString(RimWithTyre.RimBoltDistance.Value) != string.Empty)
                RimTags.Add("Болт разстояние", ConverterToString.RimBoltDistanceString(RimWithTyre.RimBoltDistance.Value));

            if (RimWithTyre.RimBoltDistance != null && ConverterToString.BoltDistanceString(RimWithTyre.RimBoltDistance.Value) != string.Empty)
                RimTags.Add("Болт дистанция", ConverterToString.BoltDistanceString(RimWithTyre.RimBoltDistance.Value));

            if (only)
                NormalzeCommon();
        }

        private void NormalizeRimWithTyre()
        {
            NormalizeRim(false);
            NormalizeTyre(false);
            NormalzeCommon();
        }

        private void NormalizePartCar()
        {
            if (Part.EngineModel != null && Part.EngineType.Value != 0)
                EngineTypeDesc = ConverterToString.EngineTypeStr(Part.EngineType.Value);
            if (Part.GearboxType != null && Part.GearboxType.Value != 0)
                GearboxDesc = ConverterToString.GearboxTypeStr(Part.GearboxType.Value);
            if (Part.GearboxType != null && Part.GearboxType.Value != 0 && ConverterToString.GearboxTypeStr(Part.GearboxType.Value) != string.Empty)
                Tags.Add("Скоростна кутия", ConverterToString.GearboxTypeStr(Part.GearboxType.Value));

            if (Part.EngineType != null && Part.EngineType.Value != 0 && ConverterToString.EngineTypeStr(Part.EngineType.Value) != string.Empty)
                Tags.Add("Вид двигател", ConverterToString.EngineTypeStr(Part.EngineType.Value));
            if (Part.EngineModel.Length > 0)
                Tags.Add("Код двигател", $"Код двигател:{Part.EngineModel}");

            if (!Part.IsCar)
            {
                PositionDesc = GetPosition();
                if (PositionDesc != string.Empty)
                    Tags.Add("Позиция", $"Позиция {PositionDesc}");
            }
            if (Part.PowerkWh != null && Part.PowerkWh != 0)
                Tags.Add("Мощност", $"{Part.PowerkWh}KW/{Part.PowerBHP}BHP ");
            if (Part.Vin?.Length > 0)
                Tags.Add("VIN", $"{Part.Vin}");
            if ((ItemType == ItemType.BusPart || ItemType == ItemType.CarPart)  && Part.PartNumber.Length > 0 )
                Tags.Add("Номер на частта", $"Номер на частта: {Part.PartNumber}");
            if (Part.Millage != 0)
                Tags.Add("Пробег", $"{Part.Millage} км");


        }
        public void NormalzeCommon()
        {
            if (ItemType == ItemType.CarPart || ItemType == ItemType.BusPart || ItemType == ItemType.OnlyBus || ItemType == ItemType.OnlyCar) {
                if (ItemType == ItemType.OnlyCar)
                    DescriptionModel = $"Колa на части - {Part.CompanyName} {Part.ModelName}    {Part.ModificationName}";
                else if (ItemType == ItemType.OnlyBus)
                    DescriptionModel = $"Бус на части - {Part.CompanyName} {Part.ModelName}";
                else
                {
                    DescriptionModel = $"{Part.DealerSubCategoryName} за {Part.CompanyName} {Part.ModelName} {Part.ModificationName}";
                }
            } 
            else
            {
                if (ItemType == ItemType.Rim)
                {
                    DescriptionModel = "Джанта";
                }
                else if (ItemType == ItemType.Tyre)
                {
                    DescriptionModel = "Гума";
                }
                else if (ItemType == ItemType.RimWithTyre)
                {
                    DescriptionModel = "Гума с джанта";
                }
            }

            if (MainPicture == null)
                MainPicture = "assets/NoImage.jpg";

            if (ConverterToString.RegionString(RegionId) != string.Empty)
                Tags.TryAdd("Регион", ConverterToString.RegionString(RegionId));

            if (Description?.Length > 0)
                Tags.Add("Описание", $"{Description}");
        }
    }
}
